### ISSUE TITLE
Khala M6: learned coordinator (TRINITY lane), shadow-deployed

### ISSUE THREAD

**Issue by AtlantisPleb** [OpenAgentsInc/openagents#6014]
**Khala** = our OpenAI-compatible inference model/gateway (`docs/inference/khala.md`). NOTE: the earlier "Khala" websocket sync-engine usage (`area:khala` / `khala-*` labels) is **deprecated/superseded** by this. Part of the Khala inference-model buildout (see master tracking issue). Source: [`docs/inference/khala-buildout-roadmap.md`](https://github.com/OpenAgentsInc/openagents/blob/main/docs/inference/khala-buildout-roadmap.md).

## M6 — Learned coordinator (TRINITY lane), shadow-deployed

**Workstream:** A (serving) + C (training)

**Goal.** Replace the heuristic router with a tiny learned coordinator, trained by evolution against the verification verdict, shipped as a shadow candidate.

**Scope / tasks.**
- [ ] Build Psionic coordinator primitives P1–P5: P1 hidden-state extraction (`forward_with_hidden`), P2 coordinator head + frozen backbone (+ optional SVF), P3 sep-CMA-ES optimizer, P4 scalar terminal-reward adapter + atomic-eval harness, P5 typed worker-pool binding.
- [ ] Train the logit router (pick worker + role) by sep-CMA-ES; reward = the verification verdict (train on verdict, monetize on settlement).
- [ ] Ship as a **shadow candidate** vs the heuristic router under Psionic's promoted/candidate contract; compare on cost-per-accepted-outcome (verified-work-per-sat).
- [ ] Promotion is a gated `runtime_promotion` (approval-gated, same governance as Artanis authority).

**Done when.** The learned router beats the heuristic router on cost-per-accepted-outcome in shadow on the in-distribution task set.

**Depends on.** M2 (verdict as reward), M4 (real pool). 

**References.** `docs/sakana/psionic-coordinator-roadmap.md` (P1–P5); `docs/sakana/coordinator-as-verified-work.md`; `docs/sakana/trinity.md`.

**Comment by AtlantisPleb** [comment]
## M6 substrate complete — P1–P5 merged (2026-06-22)

The TRINITY coordinator substrate is fully landed on `psionic`:
- **P1+P2** (psionic #1133, `d99eb130`): `forward_with_hidden` hidden-state extraction + `CoordinatorHead`.
- **P3–P5** (psionic #1134, `2ea46985`): separable CMA-ES optimizer (gradient-free, via the `flatten_parameters`/`with_flat_parameters` seam) + `TerminalRewardAdapter`/`evaluate_coordinator` hook + capability-filtered `WorkerPoolBinding`. CPU smoke drives a real head from a failing verdict (0.0) to verified (1.0); 11/11 module tests; `RandomSearch` baseline beaten at equal budget.

A dry-run **GCloud** job spec (CPU-only `e2-standard-4`, bounded self-deleting VM, provisions nothing without `--submit`) and a **Pylon-served eval** fan-out note shipped with it.

**Remaining for M6:** the actual training **run** — live `CoordinatorFitness` over real rollouts, `forward_with_hidden` on a frozen Qwen3-0.6B, Pylon fan-out with a per-generation sat budget, then shadow-ship the evolved head as a `Candidate`. The optimizer is CPU-cheap; the real cost is per-eval worker spend (budget before launch). Substrate is done; this is now a compute/budget decision, not missing code.

**Comment by AtlantisPleb** [comment]
M6 live-training lane merged (psionic#1135). sep-CMA-ES drives a **real** `CoordinatorHead` (reading real frozen-backbone `forward_with_hidden` hidden states) from fitness 0.333→1.0 over 8001 evals, routing samples to capability-eligible workers — labeled a **smoke** (deterministic verdict source), **0 sats spent**. `DailySpendCap` clamps to the shared `OWNER_DAILY_CAP_MSATS` (10,000 sats/day; mirrors the buy-mode `daily_cap_msats`/`spent_today_msats`/`day_key` ceiling, fails closed). **A real sat-moving run is held** pending a reachable live Pylon `EvalVerdictSource` + the Tassadar `training.verification_classes.v1` verdict + a spend-enabled buy-mode campaign row to debit the shared cap. Substrate (P1–P5) + the live-training driver are now all merged.

**Comment by AtlantisPleb** [comment]
Closeout pass note (2026-06-23): not safe to close yet. Psionic P1-P5, the live-training driver, and capped preflight are merged, but the M6 done condition still needs a real sat-moving shadow/deployment run over live rollouts and a reachable live Pylon/verdict source. This is tracked in workspace NEEDS_OWNER.md as a live evidence/spend gate under the existing owner-paid-run authorization and caps.

**Comment by AtlantisPleb** [comment]
Claiming a narrow OpenAgents/Pylon slice now: M6 shadow-run readiness preflight projection. Scope is public-safe/read-only code that composes existing Psionic training boundary evidence, real Pylon serving preflight evidence, owner approval/spend-cap inputs, live verdict-source refs, shadow candidate refs, and paid-shadow-win refs into an explicit fail-closed readiness object. No Psionic runtime work, no M7/M8 work, no dispatch, no live spend, no real GPU adapter execution, and I will not close this issue unless the remaining real sat-moving shadow/deployment run evidence exists. Coordination claim is recorded in `2026-06-23-khala-continuation-delegation-audit.md` on workspace main commit `ae8bc67`.

**Comment by AtlantisPleb** [comment]
Implemented the claimed OpenAgents/Pylon M6 shadow-run readiness projection on `main` commit `8ea8550d21`.

Summary:
- Added `apps/pylon/src/khala-m6-shadow-preflight.ts` with schema `openagents.khala.m6.shadow_run_preflight.v0.1`.
- The projection is read-only/public-safe and composes owner confirmation + approval ref, capped spend inputs, Psionic training-boundary evidence, real Pylon serving preflight evidence, live verdict source, shadow candidate, baseline router, live rollout, paid-shadow-win, and publication refs.
- It distinguishes `canStartShadowRun` from `canPublishM6Claim`: a capped shadow run can become start-ready before the paid-shadow/publication refs exist, but the public M6 claim cannot green without them.
- Added focused tests and docs in `apps/pylon/docs/serving-capability-evidence.md` and `docs/inference/khala-buildout-roadmap.md`.

Verification:
- `bun test apps/pylon/tests/khala-m6-shadow-preflight.test.ts apps/pylon/tests/psionic-training-boundary.test.ts apps/pylon/tests/serving-capability.test.ts` passed: 31 tests.
- `bun run --cwd apps/pylon typecheck` passed.
- `git diff --check` passed.
- Full pre-push `check:deploy` passed and allowed the push.

Not closing this issue: this did not run Psionic, dispatch live work, spend sats, arm a Pylon, execute the real shadow/deployment run, or produce a paid shadow-win receipt. Remaining close gate is still the real sat-moving M6 shadow/deployment run over live rollouts with reachable live Pylon/verdict source and publication refs.

**Comment by AtlantisPleb** [comment]
Claiming the next narrow #6014 / Psionic M6 bridge now: add a default-off typed HTTP implementation of the Psionic `BuyModeDispatch` seam, so an owner-armed OpenAgents Worker/buy-mode eval endpoint can become the live `EvalVerdictSource` later. Scope is Psionic code/tests/docs only; no live dispatch, no spend, no OpenAgents Worker settlement/payout changes, no M7/M8 reducer work, no Pylon route/secrets, no MPP/Stripe, no deploy, and I will not close this issue unless a real sat-moving M6 shadow/deployment run over live rollouts exists. Coordination claim is pushed at AtlantisPleb/workspace@8c3308f.

**Comment by AtlantisPleb** [comment]
Implemented the claimed Psionic M6 default-off HTTP buy-mode dispatch bridge in psionic commit 59210568 (pushed to OpenAgentsInc/psionic main).\n\nSummary:\n- Added `HttpBuyModeDispatch`, a `BuyModeDispatch` implementation for an owner-armed OpenAgents Worker HTTP endpoint.\n- Added fail-closed env parsing: the lane only configures when `PSIONIC_BUY_MODE_HTTP_ARM=armed`; endpoint/token are required after arming; non-http(s), URL credentials, missing host, zero timeout, and malformed arm values fail closed.\n- Redacted endpoint/token/body from config debug and transport errors.\n- Added local loopback HTTP tests for request shape, bearer auth, successful verdict decode, status errors, and schema errors. No live endpoint was called and no sats moved.\n- Updated Psionic M6 docs to distinguish the existing signed TCP dispatcher from this default-off Worker HTTP bridge.\n\nVerification:\n- `git diff --check`\n- `cargo test -p psionic-train --lib coordinator_http_buymode_dispatch -- --nocapture` (7 passed)\n- `cargo test -p psionic-train --lib coordinator_live_buymode_dispatch -- --nocapture` (14 passed)\n- `cargo test -p psionic-train --lib coordinator_eval_verdict_source -- --nocapture` (11 passed)\n- `cargo test -p psionic-train --lib coordinator_ -- --nocapture` (109 passed)\n\nI am not closing #6014: the remaining close gate is still an actual owner-armed, sat-moving M6 shadow/deployment run against the real Pylon pool + spend-enabled buy-mode campaign row, with public-safe evidence.

**Comment by AtlantisPleb** [comment]
Claiming the next narrow #6014 Worker-side bridge: add the OpenAgents HTTP buy-mode eval adapter endpoint that the Psionic `HttpBuyModeDispatch` client expects. Scope is admin/operator-authenticated Worker code/tests/docs that decode the Psionic `BuyModeEvalJob` shape and return compact `{ verdict, settled_msats }` results through existing buy-mode boundaries. No live M6 run, no production spend arming, no deploy/secrets, no M7/M8, no Pylon DNS/tunnel, no MPP/Stripe, and I will not close #6014 unless the actual sat-moving shadow/deployment run exists. Coordination claim is pushed at AtlantisPleb/workspace@d9688be.

**Comment by AtlantisPleb** [comment]
Implemented the claimed Worker-side Psionic M6 HTTP buy-mode eval adapter on OpenAgents main commit bf179fa9c7.\n\nSummary:\n- Added POST /api/operator/buy-mode/eval as the admin/operator-authenticated HTTP target for Psionic's HttpBuyModeDispatch client.\n- Decodes Psionic's compact BuyModeEvalJob wire shape: worker_id, role_index, sample_id, amount_msats.\n- Derives stable idempotency from the job, then admits through the existing buy-mode campaign/per-job-cap/daily-cap/relay dispatch boundary before any eval result can be returned.\n- Returns Psionic's expected compact response shape: { verdict: { class, passed }, settled_msats }.\n- Fails closed with blocker.buy_mode.eval_bridge_unconfigured unless a live eval bridge is configured in the Worker runtime; cap breaches halt before relay publication or eval-bridge calls.\n- Updated apps/pylon/docs/serving-capability-evidence.md to document the endpoint and the remaining runtime gate.\n\nVerification:\n- bun run --cwd apps/openagents.com/workers/api test -- src/operator-buy-mode-routes.test.ts (9 passed)\n- bun run --cwd apps/openagents.com/workers/api typecheck\n- git diff --check\n- bun run --cwd apps/openagents.com check:deploy\n- post-rebase focused route tests\n- normal pre-push check:deploy hook passed before main push\n\nNot closing #6014: this did not configure the live eval bridge, deploy Worker secrets/flags, run Psionic, arm spend, spend sats, execute the M6 shadow/deployment run, or produce paid-shadow/publication refs. The remaining close gate is still the actual owner-armed sat-moving M6 shadow/deployment run over live rollouts against the real Pylon pool + spend-enabled buy-mode campaign, with the live eval bridge configured and public-safe evidence refs.

**Comment by AtlantisPleb** [comment]
M6 closeout complete.

Summary:
- Deployed the OpenAgents buy-mode live relay/eval/settlement bridge and wired Pylon wallet/provider support for the bounded paid shadow. Latest production Worker deploy used for the run: b821cab6-81fd-4947-816d-5268f9faeafe.
- Ran `coordinator_live_train -- --real` from Psionic against campaign `buy_mode_campaign_khala_m6_issue6014_20260624160649` and live Pylon NIP-90 providers.
- Receipt schema: `psionic.khala_m6.paid_shadow_run.v1`.
- Result: learned lane 3/3 verified at 3 sats total vs heuristic lane 3/3 verified at 6 sats total.
- Metric: learned verified-work-per-sat 1.0 vs heuristic 0.5.
- Spend: 9000 msats under the local 10000 msat run cap.
- Decision: `promotion_eligible` / `promote_candidate`. Runtime promotion remains approval-gated and was not automatically flipped.

Public settlement refs:
- learned: `receipt.public.buy_mode.m6.50b9e65f129f0b16f3eceebe`, `settlement.public.buy_mode.m6.28243b5ec404c27381c00fbf`, `receipt.public.buy_mode.m6.57d6cfeb442a1460a917c18f`, `settlement.public.buy_mode.m6.2ea92e2f9093bd8751c4832c`, `receipt.public.buy_mode.m6.2276621c58367e594fac34ba`, `settlement.public.buy_mode.m6.84b51fdd3bba5f1bfc394dc2`
- heuristic: `receipt.public.buy_mode.m6.6fee0502464ab026954920fd`, `settlement.public.buy_mode.m6.00840ae39783b2484a0a237e`, `receipt.public.buy_mode.m6.15f656359089febb3e119ba1`, `settlement.public.buy_mode.m6.cc1905e459718b99e7bd55d9`, `receipt.public.buy_mode.m6.e81a389a627ce4538e036d0f`, `settlement.public.buy_mode.m6.be40b84484d9eb19fa405b88`

Docs/audit updated:
- `docs/launch/2026-06-24-khala-m6-paid-shadow-run.md`
- Psionic coordinator docs updated for the live closeout.
- Workspace continuation audit updated.

Commits pushed to main:
- OpenAgents: `125685362f`
- Psionic: `a7f65c9568`
- Workspace audit: `4a53df9eb2`

Verification:
- OpenAgents `bun run check:deploy`: green, including pre-push gate.
- Worker focused tests: 18 passed.
- Pylon provider/wallet/autostart tests: 49 passed.
- Psionic `cargo check -p psionic-train --bin coordinator_live_train`: passed with existing warnings.
- Live paid shadow run completed with public-safe receipt refs above.

**Comment by AtlantisPleb** [comment]
Closing as complete: bounded live paid shadow is done, docs/audit are updated, commits are pushed to main, and the learned coordinator beat the heuristic on verified work per sat in the live shadow run.
